### PR TITLE
🎉New Source: TMDb [low-code cdk]

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
Developing new connector for the source: TMDb

Resolves
[🎉 New Source: TMDb](https://github.com/airbytehq/connector-contest/issues/137)

## How
Developed using low-code cdk

## Streams
- Movies_alternative_titles
- Movies_changes
- Movies_credits
- Movies_details
- Movies_external_ids
- Movies_images
- Movies_keywords
- Movies_latest
- Movies_lists
- Movies_now_playing
- Movies_popular
- Movies_recommentations
- Movies_releases_dates
- Movies_reviews
- Movies_similar_movies
- Movies_top_rated
- Movies_translations
- Movies_upcoming
- Movies_videos
- Movies_watch_providers

## Recommended reading order
1. spec.yaml
2. tmdb.yaml
3. schemas/*

## 🚨 User Impact 🚨
 - No breaking changes, Just addition of new source

<details><summary><strong>New Connector: TMDb</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

## Tests

<details><summary><strong>Integration & Acceptance</strong></summary>

Result: Passed 
Run ID: `5f6361892718fd5a37bf53cd76049397cdbaa531`

image

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Hello @btkcodedev thanks for the contribution! Please request my review when the PR is ready.

**Comment by btkcodedev** [comment]
@marcosmarxm It's ready for review

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 1]
```diff
@@ -0,0 +1,189 @@
+version: "0.1.0"
+
+definitions:
+  selector:
+    extractor:
+      field_pointer: []
+  requester:
+    url_base: "https://api.themoviedb.org/3/"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:8]: Can you implement only the latets version (v4): https://developers.themoviedb.org/4/getting-started/authorization
and remove all the api_key requests in the streams.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+documentationUrl: https://docsurl.com
+connectionSpecification:
+  $schema: http://json-schema.org/draft-07/schema#
+  title: Tmdb Spec
+  type: object
+  required:
+    - api_key
+    - movie_id
+  additionalProperties: true
+  properties:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:10]: The API support language, are you planning to add this parameter to filter movies?

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    api_key:
+      type: string
+      description: API Key
+      airbyte_secret: true
+    movie_id:
+      type: string
+      description: Target movie ID
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:17]: I don't think this is the best usage of the connector. What about using the search endpoint to retrieve all movies after a release date?

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+documentationUrl: https://docsurl.com
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:1]: Update this, it must be the airbyte.io doc to your connector. See other connector to get one example.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments!

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 1 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:8]: But the v4 doesn't support movies. Also it is in development stage, thus will have bugs right?

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:8]: Agree, so focus in using the auth for v3.

File: airbyte-integrations/connectors/source-tmdb/integration_tests/catalog.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+  "streams": [
+    {
+      "name": "TODO fix this file",
+      "supported_sync_modes": ["full_refresh", "incremental"],
+      "source_defined_cursor": true,
+      "default_cursor_field": "column1",
+      "json_schema": {
+        "$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/integration_tests/catalog.json:9]: Remove this file.

File: airbyte-integrations/connectors/source-tmdb/integration_tests/sample_config.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+  "fix-me": "TODO"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/integration_tests/sample_config.json:3]: Update this file.

File: airbyte-integrations/connectors/source-tmdb/integration_tests/invalid_config.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "todo-wrong-field": "this should be an incomplete config file, used in standard tests"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/integration_tests/invalid_config.json:3]: update this too

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by marcosmarxm** [comment]
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml (hunk 3 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml]: Comment Resolved: Changed the flow of streams and made movie_id as optional.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml (hunk 4 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml]: Updated!

File: airbyte-integrations/connectors/source-tmdb/integration_tests/catalog.json (hunk 5 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/integration_tests/catalog.json]: Done!

File: airbyte-integrations/connectors/source-tmdb/integration_tests/sample_config.json (hunk 6 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/integration_tests/sample_config.json]: Done!

File: airbyte-integrations/connectors/source-tmdb/integration_tests/invalid_config.json (hunk 7 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/integration_tests/invalid_config.json]: Done!

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:9]: Comment resolved: Added language in the stream path

**Comment by btkcodedev** [comment]
@marcosmarxm @harshithmullapudi All comments resolved.
Review from community needed.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:9]: Added config

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+      type: JsonSchema
+      file_path: "./source_tmdb/schemas/{{ options['name'] }}.json"
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+
+  certification_movie_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "certification_movie"
+      path: "/certification/movie/list?api_key={{ config['api_key'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:32]: @btkcodedev why are you adding the api_key here and in the authentication? Keep only the Auth method.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "changes_person"
+      path: "/person/changes?api_key={{ config['api_key'] }}&page=1"
+
+  movies_details_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "movies_details"
+      path: "/movie/{{ config['movie_id'] or '550' }}?api_key={{ config['api_key'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:62]: I don't think this a good way, make the `movie_id` required and remove the default value.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "trending"
+      path: "/trending/all/day?api_key={{ config['api_key'] }}"
+
+  search_companies_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "search_companies"
+      path: "/search/company?query={{ config['api_key'] or 'Marvel'}}&api_key={{ config['api_key'] }}&page=1"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:188]: Same here, remove the Marvel.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments, but the connector is working! :) Just some final details.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 8 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml]: Done

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 9 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml]: Done

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 10 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml]: Done

**Comment by btkcodedev** [comment]
@marcosmarxm All comments resolved. Thanks!

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    url_base: "https://api.themoviedb.org/3/"
+    http_method: "GET"
+    request_options_provider:
+      request_parameters:
+        api_key: "{{ config['api_key'] }}"
+  retriever:
+    record_selector:
+      $ref: "*ref(definitions.selector)"
+    paginator:
+      type: NoPagination
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:17]: @btkcodedev you must implement Pagination here. you're only retrieving the first page data.
Check IncrementalPaginator: https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/pagination#page-increment

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "search_companies"
+      path: "/search/company?query={{ config['query'] }}"
+
+  search_collections_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "search_collections"
+      path: "/search/collection?query={{ config['query'] }}&language={{ config['language'] or 'en-US' }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:193]: Move the default value to spec.yaml

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+      type: string
+      description: Target movie ID, Mandate for movie streams (Example is 550)
+    query:
+      title: Optional Query for search streams
+      type: string
+      description: Target movie ID, Mandate for search streams (Example is Marvel)
+    language:
+      title: Optional language for filtering
+      type: string
+      description: Language expressed in ISO 639-1 scheme, Mandate for required streams (Example en-US)
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:28]: Add default value here.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+      description: API Key from tmdb account
+      airbyte_secret: true
+    movie_id:
+      title: Optional movie ID for targeting movies
+      type: string
+      description: Target movie ID, Mandate for movie streams (Example is 550)
+    query:
+      title: Optional Query for search streams
+      type: string
+      description: Target movie ID, Mandate for search streams (Example is Marvel)
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:24]:
```suggestion
    query:
      title: Optional Query for search streams
      type: string
      description: Target movie ID, Mandate for search streams
      examples: 
        - Marvel
```

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml:24]: Check other parameter and correct them.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments.

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 11 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml:39]: Done

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml (hunk 12 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/tmdb.yaml]: Done

File: airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml (hunk 13 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-tmdb/source_tmdb/spec.yaml]: Done

**Comment by btkcodedev** [comment]
@marcosmarxm All comments resolved and tested.

**Comment by marcosmarxm** [comment]
@btkcodedev can you share the output of `./gradlew airbyte-integrations:connectors:source-tmdb:integrationTest` I'm running with config:
```json
{
    "api_key": "my api key",
    "movie_id": "550",
    "query": "Marvel"
  }
```
and it isn't working.

**Comment by btkcodedev** [comment]
> @btkcodedev can you share the output of `./gradlew airbyte-integrations:connectors:source-tmdb:integrationTest` I'm running with config:
> 
> ```json
> {
>     "api_key": "my api key",
>     "movie_id": "550",
>     "query": "Marvel"
>   }
> ```
> 
> and it isn't working.

Oo my bad, I didn't add language to the required spec, But was added to my secret config.
No problem, fixed it, Just add language: en-US in config. It will work
Recent test resulted in a pass state.

image

**Comment by btkcodedev** [comment]
@marcosmarxm My config:
{
  "api_key": "API_KEY",
  "movie_id": "550",
  "query": "Marvel",
  "language": "en-US"
}

And all comments resolved! :)

**Comment by btkcodedev** [comment]
@marcosmarxm Could somebody please review my PR, I've worked so hard to get theseuch of streams right. Please get it merged before deadline

**Comment by btkcodedev** [comment]
Anyonee???

**Comment by marcosmarxm** [comment]
@bthecoderdev I'm review today. Don't worry about the deadline.

**Review by marcosmarxm — APPROVED** [review]
Thanks @btkcodedev

**Pull request merged**